lib/pure/asyncnet.nim

```diff
@@ -111,28 +113,28 @@ const useNimNetLite = defined(nimNetLite) or defined(freertos) or defined(zephyr
 when defineSsl:
   import openssl
 
+privateAccess(SocketBaseImpl)
+privateAccess(BufferedImpl)
+
 type
   # TODO: I would prefer to just do:
   # AsyncSocket* {.borrow: `.`.} = distinct Socket. But that doesn't work.
-  AsyncSocketDesc = object
-    fd: SocketHandle
+  AsyncSocketDesc = object of SocketBaseImpl
     closed: bool     ## determines whether this socket has been closed
-    isBuffered: bool ## determines whether this socket is buffered.
-    buffer: array[0..BufferSize, char]
-    currPos: int     # current index in buffer
-    bufLen: int      # current length of buffer
-    isSsl: bool
     when defineSsl:
-      sslHandle: SslPtr
-      sslContext: SslContext
       bioIn: BIO
       bioOut: BIO
-      sslNoShutdown: bool
-    domain: Domain
-    sockType: SockType
-    protocol: Protocol
+
   AsyncSocket* = ref AsyncSocketDesc
 
+  BufferedAsyncSocketDesc = BufferedImpl[AsyncSocketDesc]
+
+  BufferedAsyncSocket = ref BufferedAsyncSocketDesc
+
+converter toBufferedAsyncSocket(socket: AsyncSocket): BufferedAsyncSocket =
```

Comment:
When will this converter be invoked? I'm not all that keen on `cast`s being done in converters, with only an assert for a guard.